### PR TITLE
[rostopic] Show stat when rostopic echo --noarr/nostr

### PR DESCRIPTION
**Pull request by wkentaro**
Should this be another option like `--statarr`?
Because this kind of function is not needed for string for me but for only array field.

screen shot 2015-12-31 at 3 17 50 am

### CODE REVIEW

File: tools/rostopic/src/rostopic/__init__.py [hunk 1]
```diff
@@ -657,11 +660,19 @@ def __init__(self, topic, msg_eval, plot=False, filter_fn=None,
...
         if type_information and type_information.startswith('uint8['):
             val = [ord(x) for x in val]
-        return genpy.message.strify_message(val, indent=indent, time_offset=time_offset, current_time=current_time, field_filter=field_filter, fixed_numeric_width=fixed_numeric_width)
+        if value_transform:
+            val, value_untransform_fn = value_transform(val)
+        text = genpy.message.strify_message(val, indent=indent, time_offset=time_offset, current_time=current_time, field_filter=field_filter, fixed_numeric_width=fixed_numeric_width)
+        if value_transform:
+            # this operation is necessary because value_transform does change the type of message
+            # and that causes not strified values next time
+            val = value_untransform_fn(val)
```
Comment (wjwwood) [inline tools/rostopic/src/rostopic/__init__.py]: This might be easier to read like this:

``` python
value_untransform_fn = None
if value_transform:
    val, value_untransform_fn = value_transform(val)
text = genpy.message.strify_message(val, indent=indent, time_offset=time_offset, current_time=current_time, field_filter=field_filter, fixed_numeric_width=fixed_numeric_width)
if value_untransform_fn is not None:
    # this operation is necessary because value_transform does change the type of message
    # and that causes not strified values next time
    val = value_untransform_fn(val)
```

**Comment by wjwwood** [comment]
Looks reasonable to me.

File: tools/rostopic/src/rostopic/__init__.py (hunk 1 above)
Comment (wkentaro) [inline tools/rostopic/src/rostopic/__init__.py]: Thanks, I updated the commit.

**Comment by wkentaro** [comment]
Thanks for the feedback, I updated the commit.

File: tools/rostopic/src/rostopic/__init__.py [hunk 2]
```diff
@@ -608,7 +608,8 @@ class CallbackEcho(object):
     def __init__(self, topic, msg_eval, plot=False, filter_fn=None,
                  echo_clear=False, echo_all_topics=False,
                  offset_time=False, count=None,
-                 field_filter_fn=None, fixed_numeric_width=None):
+                 field_filter_fn=None, value_transform_fn=None,
+                 fixed_numeric_width=None):
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py]: Please do not change the order of existing arguments but add new arguments at the end.

Same below.

File: tools/rostopic/src/rostopic/__init__.py [hunk 3]
```diff
@@ -657,11 +660,20 @@ def __init__(self, topic, msg_eval, plot=False, filter_fn=None,
...
         if type_information and type_information.startswith('uint8['):
             val = [ord(x) for x in val]
-        return genpy.message.strify_message(val, indent=indent, time_offset=time_offset, current_time=current_time, field_filter=field_filter, fixed_numeric_width=fixed_numeric_width)
+        value_untransform_fn = None
+        if value_transform is not None:
+            val, value_untransform_fn = value_transform(val)
+        text = genpy.message.strify_message(val, indent=indent, time_offset=time_offset, current_time=current_time, field_filter=field_filter, fixed_numeric_width=fixed_numeric_width)
+        if value_untransform_fn is not None:
+            # this operation is necessary because value_transform does change the type of message
+            # and that causes not strified values next time
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py]: Spelling: `strified`

File: tools/rostopic/src/rostopic/__init__.py [hunk 4]
```diff
@@ -1145,25 +1159,44 @@ def eval_fn(m):
...
+                val.__setattr__(f, '<array type: %s, length: %s>' %
+                                (t.rstrip('[]'), len(val.__getattribute__(f))))
+                index = fields.index(f)
+                val._slot_types[index] = 'string'
+                transformed_arrays.append((index, t))
             elif echo_nostr and 'string' in t:
-                continue
+                val.__setattr__(f, '<string length: %s>' % len(val.__getattribute__(f)))
+        def untransform_fn(val):
+            for index, type in transformed_arrays:
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py]: Please don't use reserved name like `type` for variable names.

File: tools/rostopic/src/rostopic/__init__.py [hunk 5]
```diff
@@ -1145,25 +1159,44 @@ def eval_fn(m):
...
+                val._slot_types[index] = type
+            return val
+        return val, untransform_fn
+    return value_transform
+
+def create_field_filter(echo_nostr, echo_noarr):
+    def field_filter(val):
+        fields = val.__slots__
+        field_types = val._slot_types
+        for f, t in zip(fields, field_types):
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py]: Why `zip` but then only use the first element?

**Comment by dirk-thomas** [comment]
Please add tests for this new command.

**Comment by dirk-thomas** [comment]
Please ignore the last comment about tests.

File: tools/rostopic/src/rostopic/__init__.py (hunk 3 above)
Comment (wkentaro) [inline tools/rostopic/src/rostopic/__init__.py]: `strified` is misspelling?

Comment (wjwwood) [inline tools/rostopic/src/rostopic/__init__.py]: He probably thought you meant `stringified`, but you meant `str`-ified

Comment (wkentaro) [inline tools/rostopic/src/rostopic/__init__.py]: I chose `strified` inspired by `custom_strify_message`. Is `stringified` better?

Comment (tfoote) [inline tools/rostopic/src/rostopic/__init__.py]: It's still very much slang but `stringified` is much better. Without that it's hard to parse the base word from the made up ending, in my eyes I first see `strife` as the base and then try to make it a verb and past tense.

**Comment by wkentaro** [comment]
I have updated `strified` -> `stringified`.

**Comment by dirk-thomas** [comment]
Some of the unit tests are currently failing. Please see the referenced Jenkins builds.

File: tools/rostopic/src/rostopic/__init__.py [hunk 6]
```diff
@@ -608,15 +608,18 @@ class CallbackEcho(object):
...
+                 value_transform_fn=None):
         """
         :param plot: if ``True``, echo in plotting-friendly format, ``bool``
         :param filter_fn: function that evaluates to ``True`` if message is to be echo'd, ``fn(topic, msg)``
         :param echo_all_topics: (optional) if ``True``, echo all messages in bag, ``bool``
         :param offset_time: (optional) if ``True``, display time as offset from current time, ``bool``
         :param count: number of messages to echo, ``None`` for infinite, ``int``
-        :param field_filter_fn: filter the fields that are strified for Messages, ``fn(Message)->iter(str)``
+        :param field_filter_fn: filter the fields that are stringified for Messages, ``fn(Message)->iter(str)``
+        :param value_transform_fn: transform the values of Messages, ``fn(Message)->Message``
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py]: Duplicate of line 622.

Comment (wkentaro) [inline tools/rostopic/src/rostopic/__init__.py]: thanks, I have fixed.

File: tools/rostopic/src/rostopic/__init__.py [hunk 7]
```diff
@@ -1145,25 +1161,46 @@ def eval_fn(m):
...
+                val.__setattr__(f, '<string length: %s>' % len(f_val))
+        def untransform_fn(val):
+            for index, type_, f, f_val in transformed:
+                val._slot_types[index] = type_
+                val.__setattr__(f, f_val)
+            return val
+        return val, untransform_fn
+    return value_transform
+
+def create_field_filter(echo_nostr, echo_noarr):
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py:1314]: I am not sure if this function is used anywhere outside by now it doesn't filter anything anymore.

**Comment by wkentaro** [comment]
Removed the field_filter function.

**Comment by dirk-thomas** [comment]
Since you have removed all unit tests for `create_field_filter` are you planning to readd them for the new function?

**Comment by wkentaro** [comment]
I'm not planning to readd them, because filtering field it not supported after this PR.
Aside from actually removing(filtering) the fields (string/array) this PR shows the stats for them.

**Comment by wkentaro** [comment]
Is this solution OK? I removed the `create_field_filter` function which will be no longer used.

**Comment by dirk-thomas** [comment]
`create_field_filter` is a public function of this Python package. While it might not be used within this file anymore other packages might still be using it and would break if it gets removed. So I don't think it should be removed since such a breakage is not necessary.

**Comment by wkentaro** [comment]
@dirk-thomas 
OK, I restored the `create_field_filter` function.

And also, I found a bug of this code with nested messages as below.
It is fixed now with the last commit.
**BEFORE**

```
% rostopic echo /raw_image_bgr/image_color --nostr --noarr
header: 
  seq: 48
  stamp: 
    secs: 1463637710
    nsecs: 575849056
  frame_id: camera
height: 1080
width: 1920
encoding: <string length: 4>
is_bigendian: 0
step: 5760
data: <array type: uint8, length: 6220800>
```

**AFTER**

```
% rostopic echo /raw_image_bgr/image_color --nostr --noarr
header: 
  seq: 46
  stamp: 
    secs: 1463637665
    nsecs: 575885057
  frame_id: <string length: 6>
height: 1080
width: 1920
encoding: <string length: 4>
is_bigendian: 0
step: 5760
data: <array type: uint8, length: 6220800>
```

**Comment by wkentaro** [comment]
@ros-pull-request-builder rostest this please

**Comment by wkentaro** [comment]
Test passed. Please review again when you have time.

**Comment by dirk-thomas** [comment]
The patch looks good. But I was thinking if the effort for transforming and then untransforming the message is necessary. Wouldn't it be easier to just copy the message and update the copied message? For efficiency you could also implement a custom copy operation which already avoids to copy the string / array fields in the first place?

**Comment by wkentaro** [comment]
> The patch looks good. But I was thinking if the effort for transforming and then untransforming the message is necessary. Wouldn't it be easier to just copy the message and update the copied message? For efficiency you could also implement a custom copy operation which already avoids to copy the string / array fields in the first place?

Thanks for the suggestion. I worked around for a while, and the problem when the transformed message was untransformed will be fixed by correct `_field_type` of the message class. So I created `TransformedMessage` as the dummy class for printing.

**Comment by wkentaro** [comment]
ping

**Comment by wkentaro** [comment]
ping

**Comment by dirk-thomas** [comment]
I am sorry for the late response but I currently have to focus on a different project and don't have much time to review ROS tickets.

The patch looks good now. Thank you for iterating on it. I cherry-picked the changed to the _kinetic-devel_ branch with two small cosmetic changes: e0f63978fe0dd239ba8002a7e79756cdf1a335d0

```
--- a/tools/rostopic/src/rostopic/__init__.py
+++ b/tools/rostopic/src/rostopic/__init__.py
@@ -831,8 +831,7 @@ class CallbackEcho(object):
             val = [ord(x) for x in val]
         if value_transform is not None:
             val = value_transform(val)
-        text = genpy.message.strify_message(val, indent=indent, time_offset=time_offset, current_time=current_time, field_filter=field_filter, fixed_numeric_width=fixed_numeric_width)
-        return text
+        return genpy.message.strify_message(val, indent=indent, time_offset=time_offset, current_time=current_time, field_filter=field_filter, fixed_numeric_width=fixed_numeric_width)

     def callback(self, data, callback_args, current_time=None):
         """
@@ -1355,7 +1354,7 @@ def create_value_transform(echo_nostr, echo_noarr):
             f_val = getattr(val, f)
             if echo_noarr and '[' in t:
                 setattr(val_trans, f, '<array type: %s, length: %s>' %
-                                (t.rstrip('[]'), len(f_val)))
+                                      (t.rstrip('[]'), len(f_val)))
                 val_trans._slot_types[index] = 'string'
             elif echo_nostr and 'string' in t:
                 setattr(val_trans, f, '<string length: %s>' % len(f_val))
```

**Comment by wkentaro** [comment]
Thanks!
